homeassistant/components/intellifire/config_flow.py

```diff
@@ -4,60 +4,90 @@
-        self, user_input: dict[str, Any] | None = None
-    ) -> FlowResult:
-        """Handle the initial step."""
-        if user_input is None:
-            return self.async_show_form(
-                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
-            )
+    def __init__(self):
+        """Initialize the Config Flow Handler."""
+        self._discovered_host: str = ""
+        self.fireplace_finder = AsyncUDPFireplaceFinder()
+        self._config_context = {}
+
+    async def _find_fireplaces(self):
+        """Perform UDP discovery."""
+        ips = await self.fireplace_finder.search_fireplace(timeout=1)
+        if ip := ips[0]:
+            self._discovered_host = ip
+
+    async def async_step_local_config(self, user_input=None):
+        """Handle local ip configuration."""
+        local_schema = vol.Schema(
+            {vol.Required(CONF_HOST, default=self._discovered_host): str}
+        )
         errors = {}
 
-        try:
-            serial = await validate_input(self.hass, user_input[CONF_HOST])
-        except (ConnectionError, ClientConnectionError):
-            errors["base"] = "cannot_connect"
-        else:
-            await self.async_set_unique_id(serial)
-            self._abort_if_unique_id_configured(
-                updates={CONF_HOST: user_input[CONF_HOST]}
-            )
-
-            return self.async_create_entry(
-                title="Fireplace",
-                data={CONF_HOST: user_input[CONF_HOST]},
-            )
+        if user_input is not None:
+            try:
+                serial = await validate_host_input(user_input[CONF_HOST])
+
+                await self.async_set_unique_id(serial)
+
+                self._config_context[CONF_HOST] = user_input[CONF_HOST]
+
+                return self.async_create_entry(
+                    title="Fireplace", data=self._config_context
+                )
+
+            except (ConnectionError, ClientConnectionError):
+                errors["base"] = "cannot_connect"
+
+                return self.async_show_form(
+                    step_id="local_config",
+                    errors=errors,
+                    description_placeholders={CONF_HOST: user_input[CONF_HOST]},
+                    data_schema=vol.Schema(
+                        {vol.Required(CONF_HOST, default=user_input[CONF_HOST]): str}
+                    ),
+                )
+
         return self.async_show_form(
-            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+            step_id="local_config", errors=errors, data_schema=local_schema
         )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Start the user flow."""
+
+        # If the integration was not triggered by DHCP attempt a quick local discovery
+        if self._discovered_host == "":
+            await self._find_fireplaces()
+
+        return await self.async_step_local_config()
```

Comment:
When we add DHCP discovery it will jump straight to `async_step_local_config` which is why I moved away from doing everything in the `async_step_user` step.